zypp/zyppng/media/network/downloader.cc

```diff
@@ -203,468 +615,851 @@ namespace zyppng {
+          }
 
-      if ( _failedBlocks.size() || ( _blockIter < _blockList.numBlocks() )) {
-        setFailed( "Unable to download all blocks." );
-        return;
-      }
+          cm.addCred( credFromUser );
+          cm.save();
 
-      if ( _state == Download::RunningMulti && _blockList.haveFileChecksum() ) {
-        //TODO move this into a external application so we do not need to block on it
-        //need to check file digest
-        zypp::Digest dig;
-        _blockList.createFileDigest( dig );
+          // potentially setting this after the file has been touched we might miss a change
+          // in a later loop, if another update to the store happens right in the few ms difference
+          // between the actual file write and calling time() here. However this is highly unlikely
+          _lastTriedAuthTime = req->_authTimestamp = time( nullptr ) ;
 
-        std::ifstream istrm( _targetPath.asString(), std::ios::binary);
-        if ( !istrm.is_open() ) {
-          setFailed( "Failed to verify file digest (Could not open target file)." );
-          return;
-        }
-        if ( !dig.update( istrm ) ) {
-          setFailed( "Failed to verify file digest (Could not read target file)." );
-          return;
-        }
-        if ( !_blockList.verifyFileDigest( dig ) ) {
-          setFailed( "Failed to verify file digest (Checksum did not match)." );
-          return;
+          retry = true;
         }
       }
-      //TODO implement digest check for non multi downloads
-      setFinished();
     }
+    return retry;
   }
 
-  void DownloadPrivate::addNewRequest( std::shared_ptr<Request> req )
+#if ENABLE_ZCHUNK_COMPRESSION
+  bool DownloadPrivateBase::hasZckInfo() const
   {
-    auto slot = _sigStarted.slots().front();
-    req->connectSignals( *this );
-    _runningRequests.push_back( req );
-    _requestDispatcher->enqueue( req );
+    if ( zypp::indeterminate(_specHasZckInfo) )
+      _specHasZckInfo = ( _spec.headerSize() > 0 && isZchunkFile( _spec.deltaFile() ) );
+    return bool(_specHasZckInfo);
   }
+#endif
 
-  std::shared_ptr<DownloadPrivate::Request> DownloadPrivate::initMultiRequest( size_t block, NetworkRequestError &err )
+  void DownloadPrivateBase::Request::disconnectSignals()
   {
-    zypp::media::MediaBlock blk = _blockList.getBlock( block );
-
-    Url myUrl;
-    TransferSettings settings;
-    if ( !findNextMirror( myUrl, settings, err ) )
-      return nullptr;
+    _sigStartedConn.disconnect();
+    _sigProgressConn.disconnect();
+    _sigFinishedConn.disconnect();
+  }
 
-    DBG << "Starting block " << block << std::endl;
+  DownloadPrivate::DownloadPrivate(Downloader &parent, std::shared_ptr<NetworkRequestDispatcher> requestDispatcher, std::shared_ptr<MirrorControl> mirrors, DownloadSpec &&spec, Download &p) :
+    DownloadPrivateBase( parent, std::move(requestDispatcher), std::move(mirrors), std::move(spec), p)
```

Comment:
CTOR initializer lists please like this
```suggestion
    : DownloadPrivateBase( parent, std::move(requestDispatcher)
    , std::move(mirrors)
    , std::move(spec), p)
```
Especially the `:` at the beginning of line not at the end of L1409